cli/tsc/mod.rs

```diff
@@ -740,8 +721,55 @@ fn op_resolve(
   Ok(resolved)
 }
 
+fn resolve_specifier_types(
+  specifier: &ModuleSpecifier,
+  state: &State,
+) -> Result<Option<(ModuleSpecifier, MediaType)>, AnyError> {
+  fn inner(
+    specifier: &ModuleSpecifier,
+    state: &State,
+    seen: &mut HashSet<ModuleSpecifier>,
+  ) -> Result<Option<(ModuleSpecifier, MediaType)>, AnyError> {
+    let graph = &state.graph;
+    Ok(match graph.get(specifier) {
+      Some(Module::Esm(module)) => {
+        if !seen.insert(module.specifier.clone()) {
+          // infinite loop, return itself
+          return Ok(Some((module.specifier.clone(), module.media_type)));
+        }
+        let maybe_types_dep = module
+          .maybe_types_dependency
+          .as_ref()
+          .map(|d| &d.dependency);
+        match maybe_types_dep.and_then(|d| d.maybe_specifier()) {
+          // follow this specifier, which might be an npm package
+          Some(specifier) => inner(specifier, state, seen)?,
+          _ => Some((module.specifier.clone(), module.media_type)),
+        }
+      }
+      Some(Module::Json(module)) => {
+        Some((module.specifier.clone(), module.media_type))
+      }
+      Some(Module::Npm(module)) => {
+        if let Some(npm_resolver) = &state.maybe_npm_resolver {
+          Some(resolve_npm_package_reference_types(
+            &module.nv_reference,
+            npm_resolver,
+          )?)
+        } else {
+          None
+        }
+      }
+      Some(Module::External(_) | Module::Node(_)) | None => None,
+    })
+  }
+
+  let mut seen = HashSet::new();
+  inner(specifier, state, &mut seen)
```

Comment:
I should add a test for this actually because it fixes a bug. Will do that tomorrow.

Reply:
Going to revert this back to the way it was.